### PR TITLE
feature/change-interpolation-method

### PR DESCRIPTION
**Pull request by nannau**
The method that we have adopted re-grids a CanRCM4 model by adding 9 times more grid cells (splitting each grid cell into 3 parts in each orthogonal axis). There have also been adjustments made to masking to soften the tolerance on what is an accepted range of coordinates.

* Adds a general re-gridding function to gridding.py
* Tests general re-gridding function
* Adds tolerances of coordinate checks in mask.py
* Uses context manager in climpyrical's datacube.read_data to comply with test dataset generated for re-gridding. This ensures that if `read_data` is called more than once in a session, the file is properly closed after it is loaded.

### CODE REVIEW

File: climpyrical/gridding.py [hunk 1]
```diff
@@ -432,26 +513,32 @@ def check_find_nearest_value_inputs(x, y, x_i, y_i, field, mask):
...
         )
     # mask same shape as field
     if field.shape != mask.shape:
         raise ValueError(
             "Field and mask are not the same shape. Received field shape \
-            {} and mask shape {}.".format(field.shape, mask.shape)
+            {} and mask shape {}.".format(
+                field.shape, mask.shape
+            )
         )
```
Comment (rod-glover) [inline climpyrical/gridding.py]: Wow, you are totally hardcore checking inputs.

File: climpyrical/mask.py [hunk 2]
```diff
@@ -77,15 +77,25 @@ def check_input_grid_coords(x, y):
...
+
+    if (not np.isclose(x.max(), 33.8800048828125, atol=xtol)) or (
+        not np.isclose(x.min(), -33.8800048828125, atol=xtol)
     ):
         raise ValueError("Unexpected range of values in x dimension")
-    if (not np.isclose(y.max(), 28.15999984741211)) or (
-        not np.isclose(y.min(), -28.59999656677246)
+    if (not np.isclose(y.max(), 28.15999984741211, atol=ytol)) or (
+        not np.isclose(y.min(), -28.59999656677246, atol=ytol)
     ):
```
Comment (rod-glover) [inline climpyrical/mask.py]: Possibly this answers my question about computing the mean grid cell size.

File: climpyrical/tests/test_gridding.py [hunk 3]
```diff
@@ -50,6 +57,32 @@ def test_check_ndims(data, n, error):
...
+    if error is None:
+        check_regrid_ensemble_inputs(ds, dv, n, keys)
+    else:
+        with pytest.raises(error):
+            check_regrid_ensemble_inputs(ds, dv, n, keys)
+
+
+@pytest.mark.parametrize(
+    "ds,dv,n,keys", [(ds, dv, 3, {"rlat", "rlon", "lat", "lon", "level"})]
+)
```
Comment (rod-glover) [inline climpyrical/tests/test_gridding.py]: Do you plan to have other cases? I see a mix here of parametrization on args to `regrid_ensemble`, but comparing to a fixed result (`ds_regridded_proper`). That is not often done. Usually it's all one way (params) or the other (fixed values).

File: climpyrical/tests/test_gridding.py [hunk 4]
```diff
@@ -325,10 +345,11 @@ def test_check_find_nearest_value_inputs(
...
-        or final.size != y_i.size
-        or (np.allclose(expected, final) is False)
+            np.any(np.isnan(final))
+            or final.size != x_i.size
+            or final.size != y_i.size
+            or np.allclose(expected, final)
+            is False
     )
 
     assert truth is False
```
Comment (rod-glover) [inline climpyrical/tests/test_gridding.py]: Now that statement is designed to give any philosopher a headache.

File: climpyrical/gridding.py [hunk 5]
```diff
@@ -88,8 +87,84 @@ def check_input_coords(x, y, ds):
...
+
+    # define new boundaries
+    x1 = ds.rlon.min() - dx
+    x2 = ds.rlon.max() + dx
+    y1 = ds.rlon.min() - dy
+    y2 = ds.rlat.max() + dy
+
+    # define new coordinate arrays
+    new_x = np.linspace(x1, x2, ds.rlon.size * n)
+    new_y = np.linspace(y1, y2, ds.rlat.size * n)
```
[the remark this answers is not in the document]
Comment (nannau) [reply climpyrical/gridding.py]: There is minor (extremely minor) variation along each coordinate that is not systematic. I do not know where this deviation comes from. See 
grid_size_diff_plot

So the cell sizes vary, but an extremely small amount, and I ignore that small amount by just assuming a constant, which is the mean of the differentiated array.

File: climpyrical/gridding.py [hunk 6]
```diff
@@ -88,8 +87,84 @@ def check_input_coords(x, y, ds):
...
+    check_input_coords(ds.rlon.values, ds.rlat.values, ds)
+
+
+def regrid_ensemble(
+    ds: xr.Dataset,
+    dv: str,
+    n: int,
+    keys: dict = {"rlat", "rlon", "lat", "lon", "level"},
+) -> xr.Dataset:
+    """Re-grids a regional model to have n times the
```
[the remark this answers is not in the document]
Comment (nannau) [reply climpyrical/gridding.py]: Yes, thanks.

File: climpyrical/gridding.py [hunk 7]
```diff
@@ -88,8 +87,84 @@ def check_input_coords(x, y, ds):
...
+    x2 = ds.rlon.max() + dx
+    y1 = ds.rlon.min() - dy
+    y2 = ds.rlat.max() + dy
+
+    # define new coordinate arrays
+    new_x = np.linspace(x1, x2, ds.rlon.size * n)
+    new_y = np.linspace(y1, y2, ds.rlat.size * n)
+
+    # re-create design value field on newly gridded size
+    new_ds = np.repeat(np.repeat(ds[dv].values, n, axis=1), n, axis=2)
```
[the remark this answers is not in the document]
Comment (nannau) [reply climpyrical/gridding.py]: Yes indeed.

File: climpyrical/gridding.py [hunk 8]
```diff
@@ -17,18 +19,19 @@ def check_ndims(data, n):
     """
     if not isinstance(data, np.ndarray):
         raise TypeError(
-            "Provide an array of type {}, received {}"
-            .format(np.ndarray, type(data))
+            "Provide an array of type {}, received {}".format(
+                np.ndarray, type(data)
+            )
```
[the remark this answers is not in the document]
Comment (nannau) [reply climpyrical/gridding.py]: Fancy! I will have to get in this habit. How nice.

File: climpyrical/gridding.py (hunk 1 above)
Comment (nannau) [reply climpyrical/gridding.py]: :nerd_face:

**Comment by jameshiebert** [comment]
I'll trust Rod's assessment. Feel free to proceed :)

**Pull request merged**